### PR TITLE
Await should check null-state of expression

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/31519

I'm going to wait for https://github.com/dotnet/roslyn/pull/31978 to make some improvements to this PR.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -28151,6 +28151,67 @@ class Awaiter : System.Runtime.CompilerServices.INotifyCompletion
...
+
+        [Fact]
+        public void Await_UpdateExpression()
+        {
+            var source = @"
+class C
+{
+    async void M(System.Threading.Tasks.Task<string>? task)
+    {
+        await task; // warn
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: worth testing `await x?.M()`?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Yes, I'll add. 
I expect this produces a safety warning for awaiting something that could be `null`, but that we don't update the state of `x` to non-null (if the await didn't throw). So there will be cascading diagnostics, which we could suppress.

**Comment by 333fred** [comment]
Worth adding tests for nullable of ValueTask as well?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -4901,10 +4902,16 @@ private void CheckPossibleNullReceiver(BoundExpression receiverOpt, bool checkNu
                         return;
                     }
                     ReportSafetyDiagnostic(isValueType ? ErrorCode.WRN_NullableValueTypeMayBeNull : ErrorCode.WRN_NullReferenceReceiver, syntaxOpt ?? receiverOpt.Syntax);
-                    int slot = MakeSlot(receiverOpt);
-                    if (slot > 0)
+
+                    if (receiverOpt != null)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: No need for this check, we're already in an if with this condition.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -4491,6 +4491,7 @@ private TypeSymbolWithAnnotations InferResultNullabilityOfBinaryLogicalOperator(
         public override BoundNode VisitAwaitExpression(BoundAwaitExpression node)
         {
             var result = base.VisitAwaitExpression(node);
+            CheckPossibleNullReceiver(node.Expression);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: What about the case where `GetAwaiter` is an extension method which takes a nullable receiver? What should the code do in that case and is it handled / tested here?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I'll add a test, but I expect the behavior is the same as for invocations. Namely, we still warn. That was an explicit decision in a previous PR, but we could adjust later.

**Pull request merged**